fem/particleset.cpp

```diff
@@ -0,0 +1,950 @@
+      pv.SetNumParticles(new_np); // does not delete existing data
+   }
+
+   // Update tags
+   for (int t = 0; t < GetNTags(); t++)
+   {
+      tags[t]->SetSize(new_np);
+   }
+}
+
+#if defined(MFEM_USE_MPI) && defined(MFEM_USE_GSLIB)
+
+/// \cond DO_NOT_DOCUMENT
+template<size_t NBytes>
+void ParticleSet::TransferParticlesImpl(ParticleSet &pset,
+                                        const Array<unsigned int> &send_idxs, const Array<unsigned int> &send_ranks)
+{
+   struct pdata_t
+   {
+      alignas(real_t) std::array<std::byte, NBytes> data;
+      IDType id;
+   };
+
+   int nreals = pset.GetFieldVDims().Sum() + pset.Coords().GetVDim();
+   int ntags = pset.GetNTags();
+   // in-case particles have not been initialized on all ranks
+   MPI_Allreduce(MPI_IN_PLACE, &nreals, 1, MPI_INT, MPI_MAX, pset.GetComm());
+   MPI_Allreduce(MPI_IN_PLACE, &ntags, 1, MPI_INT, MPI_MAX, pset.GetComm());
+   size_t nbytes = nreals*sizeof(real_t) + ntags*sizeof(int);
+   MFEM_VERIFY(nbytes <= NBytes, "More data than can be packed.");
+
+   using parr_t = pdata_t;
+   gslib::array gsl_arr;
+   parr_t *pdata_arr;
+   array_init(parr_t, &gsl_arr, send_idxs.Size());
+   pdata_arr = (parr_t*) gsl_arr.ptr;
+
+   gsl_arr.n = send_idxs.Size();
+   for (int i = 0; i < send_idxs.Size(); i++)
+   {
+      parr_t &pdata = pdata_arr[i];
+      pdata.id = pset.GetIDs()[send_idxs[i]];
+
+      // Copy particle data directly into pdata
+      size_t counter = 0;
+      for (int f = -1; f < pset.GetNFields(); f++)
+      {
+         ParticleVector &pv = (f == -1 ? pset.Coords() : pset.Field(f));
+         for (int c = 0; c < pv.GetVDim(); c++)
+         {
+            std::memcpy(pdata.data.data() + counter, &pv(send_idxs[i], c),
+                        sizeof(real_t));
+            counter += sizeof(real_t);
+         }
+      }
+
+      // Copy tags
+      for (int t = 0; t < pset.GetNTags(); t++)
+      {
+         Array<int> &tag_arr = pset.Tag(t);
+         std::memcpy(pdata.data.data() + counter, &tag_arr[send_idxs[i]],
+                     sizeof(int));
+         counter += sizeof(int);
+      }
+   }
+
+   unsigned int nparticles = pset.GetNParticles();
+   unsigned int nsend      = send_idxs.Size();
+
+   // Transfer particles
+   sarray_transfer_ext(parr_t, &gsl_arr, send_ranks.GetData(),
+                       sizeof(unsigned int), pset.cr);
+
+   // Make sure we have enough space for received particles
+   unsigned int nrecv = gsl_arr.n;
+   int ndelete = (int)nsend - (int)nrecv;
+   if (ndelete > 0)
+   {
+      // Remove unneeded particles
+      auto datap = const_cast<unsigned int*>(send_idxs.GetData());
+      Array<unsigned int> delete_idxs(datap + nrecv,
+                                      ndelete);
+      pset.RemoveParticles(delete_idxs);
+   }
+   else
+   {
+      pset.Reserve(nparticles-ndelete);
+   }
+
+   pdata_arr = (parr_t*) gsl_arr.ptr;
+
+   // Add newly-recvd data directly to active state
+   for (int i = 0; i < nrecv; i++)
```

Comment:
```suggestion
   for (unsigned i = 0; i < nrecv; i++)
```
Signed/unsigned implicit conversion.